spring-integration-stomp/src/main/java/org/springframework/integration/stomp/outbound/StompMessageHandler.java

```diff
@@ -0,0 +1,197 @@
+	}
+
+	public void setDestinationExpression(Expression destinationExpression) {
+		Assert.notNull(destinationExpression, "'destinationExpression' must not be null.");
+		this.destinationExpression = destinationExpression;
+	}
+
+	public void setHeaderMapper(HeaderMapper<StompHeaders> headerMapper) {
+		Assert.notNull(headerMapper, "'headerMapper' must not be null.");
+		this.headerMapper = headerMapper;
+	}
+
+	@Override
+	public void setIntegrationEvaluationContext(EvaluationContext evaluationContext) {
+		this.evaluationContext = evaluationContext;
+	}
+
+	@Override
+	public void setApplicationEventPublisher(ApplicationEventPublisher applicationEventPublisher) {
+		this.applicationEventPublisher = applicationEventPublisher;
+	}
+
+	@Override
+	protected void handleMessageInternal(final Message<?> message) throws Exception {
+		if (!this.isRunning()) {
+			throw new MessageDeliveryException(message, "The StompMessageHandler [" + getComponentName() +
+					"] hasn't been connected to StompSession. Check the state of [" + this.stompSessionManager + "]");
+		}
+
+		StompHeaders stompHeaders = new StompHeaders();
+		this.headerMapper.fromHeaders(message.getHeaders(), stompHeaders);
+		if (stompHeaders.getDestination() == null) {
+			Assert.state(this.destinationExpression != null, "One of 'destination' or 'destinationExpression' must be" +
+					" provided, if message header doesn't supply 'destination' STOMP header.");
+			String destination = this.destinationExpression.getValue(this.evaluationContext, message, String.class);
+			stompHeaders.setDestination(destination);
+		}
+
+		final StompSession.Receiptable receiptable = this.stompSession.send(stompHeaders, message.getPayload());
+		if (this.stompSessionManager.isAutoReceiptEnabled()) {
+			final String destination = stompHeaders.getDestination();
+			if (this.applicationEventPublisher != null) {
+				receiptable.addReceiptTask(new Runnable() {
+
+					@Override
+					public void run() {
+						StompReceiptEvent event = new StompReceiptEvent(StompMessageHandler.this,
+								destination, receiptable.getReceiptId(), StompCommand.SEND, false);
+						event.setMessage(message);
+						applicationEventPublisher.publishEvent(event);
+					}
+
+				});
+			}
+			receiptable.addReceiptLostTask(new Runnable() {
+
+				@Override
+				public void run() {
+					if (applicationEventPublisher != null) {
+						StompReceiptEvent event = new StompReceiptEvent(StompMessageHandler.this,
+								destination, receiptable.getReceiptId(), StompCommand.SEND, true);
+						event.setMessage(message);
+						applicationEventPublisher.publishEvent(event);
+					}
+					else {
+						logger.error("The receipt [" + receiptable.getReceiptId() + "] is lost for [" +
+								message + "] on destination [" + destination + "]");
+					}
+				}
+
+			});
+		}
+	}
+
+	@Override
+	public void start() {
+		this.stompSessionManager.connect(this.sessionHandler);
```

Comment:
If you take a look to the `stompSessionManager.connect()` implementation, you'll see that it is just an *artificial* connect and we do only this:
```
public void connect(StompSessionHandler handler) {
	this.compositeStompSessionHandler.addHandler(handler);
}
```
So, any `connectIfNeeded()` doesn't make sense here. We establish `stompSession` here only within:
```
public void afterConnected(StompSession session, StompHeaders connectedHeaders) {
	StompMessageHandler.this.stompSession = session;
	StompMessageHandler.this.running = true;
}
```
Which would be possible only if the target `stompSessionManager` has been obtained the *real* connection.